File: src/db-toolkit-ui/resources/backend/_internal/core/routes/backup.py
"""Backup management routes."""

from utils.logger import logger
from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse
from typing import Optional
from datetime import datetime
from core.storage import ConnectionStorage
from core.backup_storage import BackupStorage
from core.schemas.backup import BackupCreateRequest, BackupRestoreRequest
from operations.backup_manager import BackupManager
from core.models import BackupSchedule
    

router = APIRouter()
storage = ConnectionStorage()
backup_storage = BackupStorage()
backup_manager = BackupManager()


@router.get("/backups")
async def get_backups(connection_id: Optional[str] = None):
    """Get all backups, optionally filtered by connection."""
    backups = await backup_storage.get_all_backups(connection_id)
    return {"success": True, "backups": [b.model_dump() for b in backups]}


@router.get("/backups/{backup_id}")
async def get_backup(backup_id: str):
    """Get backup by ID."""
    backup = await backup_storage.get_backup(backup_id)
    if not backup:
        raise HTTPException(status_code=404, detail="Backup not found")
    return {"success": True, "backup": backup.model_dump()}


@router.post("/backups")
async def create_backup(request: BackupCreateRequest):
    """Create new backup."""
    connection = await storage.get_connection(request.connection_id)
    if not connection:
        raise HTTPException(status_code=404, detail="Connection not found")
    
    try:
        backup = await backup_manager.create_backup(
            connection=connection,
            name=request.name,
            backup_type=request.backup_type,
            tables=request.tables,
            compress=request.compress,
        )
        return {"success": True, "backup": backup.model_dump()}
    except Exception as e:
        logger.error(f"Route error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/backups/{backup_id}/restore")
async def restore_backup(backup_id: str, request: BackupRestoreRequest):
    """Restore backup to database."""
    backup = await backup_storage.get_backup(backup_id)
    if not backup:
        raise HTTPException(status_code=404, detail="Backup not found")
    
    target_conn_id = request.target_connection_id or backup.connection_id
    target_connection = await storage.get_connection(target_conn_id)
    if not target_connection:
        raise HTTPException(status_code=404, detail="Target connection not found")
    
    try:
        await backup_manager.restore_backup(
            backup=backup,
            target_connection=target_connection,
            tables=request.tables,
        )
        return {"success": True, "message": "Backup restored successfully"}
    except Exception as e:
        logger.error(f"Route error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/backups/{backup_id}/download")
async def download_backup(backup_id: str):
    """Download backup file."""
    backup = await backup_storage.get_backup(backup_id)
    if not backup:
        raise HTTPException(status_code=404, detail="Backup not found")
    
    from pathlib import Path
    if not Path(backup.file_path).exists():
        raise HTTPException(status_code=404, detail="Backup file not found")
    
    return FileResponse(
        path=backup.file_path,
        filename=Path(backup.file_path).name,
        media_type="application/octet-stream",
    )


@router.delete("/backups/{backup_id}")
async def delete_backup(backup_id: str):
    """Delete backup."""
    success = await backup_manager.delete_backup(backup_id)
    if not success:
        raise HTTPException(status_code=404, detail="Backup not found")
    return {"success": True, "message": "Backup deleted"}


@router.post("/backups/{backup_id}/verify")
async def verify_backup(backup_id: str):
    """Verify backup integrity."""
    success = await backup_manager.verify_backup(backup_id)
    return {"success": success, "message": "Backup verified" if success else "Verification failed"}


@router.get("/backup-schedules")
async def get_schedules():
    """Get all backup schedules."""
    schedules = await backup_storage.get_all_schedules()
    return {"success": True, "schedules": [s.model_dump() for s in schedules]}


@router.post("/backup-schedules")
async def create_schedule(schedule_data: dict):
    """Create backup schedule."""
    schedule_id = f"schedule_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    next_run = backup_manager.parse_schedule(schedule_data.get('schedule', 'daily'))
    
    schedule = BackupSchedule(
        id=schedule_id,
        connection_id=schedule_data['connection_id'],
        name=schedule_data['name'],
        backup_type=schedule_data['backup_type'],
        schedule=schedule_data.get('schedule', 'daily'),
        tables=schedule_data.get('tables'),
        compressed=schedule_data.get('compressed', True),
        retention_count=schedule_data.get('retention_count', 5),
        enabled=schedule_data.get('enabled', True),
        next_run=next_run.isoformat() if next_run else None,
    )
    
    await backup_storage.add_schedule(schedule)
    return {"success": True, "schedule": schedule.model_dump()}


@router.put("/backup-schedules/{schedule_id}")
async def update_schedule(schedule_id: str, updates: dict):
    """Update backup schedule."""
    schedule = await backup_storage.update_schedule(schedule_id, **updates)
    if not schedule:
        raise HTTPException(status_code=404, detail="Schedule not found")
    return {"success": True, "schedule": schedule.model_dump()}


@router.delete("/backup-schedules/{schedule_id}")
async def delete_schedule(schedule_id: str):
    """Delete backup schedule."""
    success = await backup_storage.delete_schedule(schedule_id)
    if not success:
        raise HTTPException(status_code=404, detail="Schedule not found")
    return {"success": True, "message": "Schedule deleted"}

File: src/db-toolkit/electron/backend/handlers/backup.js
/**
 * IPC handlers for backup operations.
 */

const { ipcMain } = require('electron');
const { BackupManager } = require('../operations/backup-manager');
const { connectionManager } = require('../utils/connection-manager');
const { connectionStorage } = require('../utils/connection-storage');
const backupStorage = require('../utils/backup-storage');

function registerBackupHandlers() {
  const backupManager = new BackupManager();

  ipcMain.handle('backup:create', async (event, data) => {
    const { connection_id, name, backup_type, tables, compress } = data;
    const connectionId = connection_id;
    const backupType = backup_type;
    
    try {
      const config = await connectionStorage.getConnection(connectionId);
      
      if (!config) {
        return { success: false, error: 'Connection not found' };
      }

      const backup = await backupManager.createBackup(null, config, name, backupType, tables, compress);
      return { success: true, backup };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:list', async (event, connectionId) => {
    try {
      const backups = await backupStorage.getAllBackups(connectionId);
      return { success: true, backups };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:get', async (event, backupId) => {
    try {
      const backup = await backupStorage.getBackup(backupId);
      return { success: true, backup };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:restore', async (event, backupId, targetConnectionId) => {
    try {
      const backup = await backupStorage.getBackup(backupId);
      const targetConnection = await connectionManager.getConnector(targetConnectionId);
      const targetConfig = await connectionStorage.getConnection(targetConnectionId);
      
      if (!backup || !targetConnection || !targetConfig) {
        return { success: false, error: 'Backup or connection not found' };
      }

      await backupManager.restoreBackup(backup, targetConnection, targetConfig);
      return { success: true };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:delete', async (event, backupId) => {
    try {
      const result = await backupManager.deleteBackup(backupId);
      return { success: result };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:verify', async (event, backupId) => {
    try {
      const result = await backupManager.verifyBackup(backupId);
      return { success: true, verified: result };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:schedule:create', async (event, schedule) => {
    try {
      const scheduleId = `schedule_${Date.now()}`;
      const now = new Date();
      const { backupScheduler } = require('../operations/backup/backup-scheduler');
      
      const nextRun = backupScheduler.calculateNextRun(schedule.schedule, now);
      
      const scheduleData = {
        ...schedule,
        id: scheduleId,
        created_at: now.toISOString(),
        next_run: nextRun.toISOString(),
        last_run: null
      };
      
      const result = await backupStorage.addSchedule(scheduleData);
      return { success: true, schedule: result };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:schedule:list', async (event) => {
    try {
      const schedules = await backupStorage.getAllSchedules();
      return { success: true, schedules };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:schedule:update', async (event, scheduleId, updates) => {
    try {
      const result = await backupStorage.updateSchedule(scheduleId, updates);
      return { success: true, schedule: result };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:schedule:delete', async (event, scheduleId) => {
    try {
      const result = await backupStorage.deleteSchedule(scheduleId);
      return { success: result };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:get-all', async (event, connectionId) => {
    try {
      const backups = await backupStorage.getAllBackups(connectionId);
      return { success: true, backups };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('backup:download', async (event, backupId) => {
    const { logger } = require('../utils/logger');
    const fs = require('fs').promises;
    const path = require('path');
    const os = require('os');
    
    try {
      logger.info(`Download backup request: ${backupId}`);
      const backup = await backupStorage.getBackup(backupId);
      if (!backup) {
        logger.error(`Backup not found: ${backupId}`);
        return { success: false, error: 'Backup not found' };
      }
      
      const downloadsPath = path.join(os.homedir(), 'Downloads');
      const fileName = path.basename(backup.file_path);
      const downloadPath = path.join(downloadsPath, fileName);
      
      await fs.copyFile(backup.file_path, downloadPath);
      logger.info(`Backup copied to Downloads: ${downloadPath}`);
      
      return { success: true, filePath: downloadPath };
    } catch (error) {
      logger.error(`Download backup failed: ${error.message}`, error);
      return { success: false, error: error.message };
    }
  });
}

module.exports = { registerBackupHandlers };
